Use listeners for plugin integration

tiagoalvesdulce opened the pull request:
This PR implements the Observer pattern in the plugins integration by using a listener middleware on Redux.

### Solution description

To understand better why the observer pattern can be useful we need to first take a look at a plugin's basic structure:

1. **Actions**: Events that describe something that happen in the application. Each plugin has its own actions. Example: add record, change filter criteria, etc.
2. **Reducer**: It's a function that will be responsible for handling dispatched actions. It has some rules:

- They should only calculate the new state value based on the state and action arguments
- They are not allowed to change the existing state. Instead, they must make immutable updates, by copying the existing state and making changes to the copied values
- They must not do any asynchronous logic or other "side effects"

3. **State**: Plugin related data. Can be accessed by other plugins and the app-shell using selectors. Should be updated only by the plugin reducer.
4. **Thunks**: Used to write sync and async logic. It's where we fetch from the API and dispatch actions based on request results. 

So, a plugin works by dispatching actions that will be handled by the reducer and can update state. The app-shell works the same way. The difference is that it can access state and dispatch actions from all plugins and create its own as well.

Now that we know how plugins and the app-shell work, why is the observer pattern useful? **Because we can listen to actions and state changes from all our plugins and run effects when they happen.** This is great to do plugin integration. If you want to do some task every time a new record is added, for example, start listening to the add record action in the app-shell and do the work there. The records plugin continues completely isolated.

But what about creating a reducer in the app-shell that can handle the add record action? There's nothing preventing us to do this since we can have more than one reducer handling the same action. But we **can not** dispatch actions nor do side effects inside a reducer. **That's where listeners shine, we can do both there!**

**With great power comes great responsibility**. Listeners are powerful and can do basically everything so we have to create some rules:

1. Do not use the same listener for different actions. It's possible to do so with a `matcher` and a matching utility called `isAnyOf`. Please don't do that. I came up with the conclusion that using it gets in the way of debugging and readability.
2. Do not listen to actions of a plugin from another plugin. This breaks decoupling. If you are tempted to do this consider moving the logic to the app-shell instead.
3. Do not write very long listeners. Always try to break the code into smaller functions.

To read more about listeners go to the [official documentation](https://redux-toolkit.js.org/api/createListenerMiddleware)

With this change we can remove some hooks and thunks that are no longer needed. I'll do that on this PR as well.

**TODO**

- [x] Fix ticketvote dev example
- [x] Do not trigger fetchNextBatch when navigating from the details page to the Home if there are records loaded
- [x] Change details integration to use listeners
- [x] Investigate better loading and error handling when integrating multiple plugins. **Solved**: we can write util functions like `getStatusFromMultipleSlices` that receives an array of status and use array utils like `some` and `every` to return the status of a page. The same can be done to handle errors. 
- [x] Handle details error
- [x] Improve /details caching verification
- [x] Fix Infinite scroll
- [x] Write docs

victorgcramos reviewed (changes requested):
Code looking really good. Nice cleanup, example and readability!

Found 2 issues: 

1. Noticed that the scrolling behavior on Home page is not consistent, and I kinda got stuck on some infinite-loading states while performing regular scrolls on our list. You can notice that if you scroll fast enough, or if you reach the end of the first inventory page.

    Here's a preview:

    https://user-images.githubusercontent.com/22639213/171255449-9abbf425-ca6c-45d6-81da-2094096d67b4.mp4

2. "Rejected" tab is broken

    https://user-images.githubusercontent.com/22639213/171256312-0d970250-489e-4edc-97ac-48d2c3d34554.mp4





Also, left some minor inline observations.

tiagoalvesdulce commented on plugins-structure/apps/politeia/src/utils/getStatusFromMultipleSlices.js:
```diff
@@ -0,0 +1,14 @@
+export function getStatusFromMultipleSlices(statuses) {
```
Didn't handled because this util is meant to be used to get the status of multiple slices dispatched at once. This doesn't happen with inventory, since it is the one controlling the flow. Let me know if you have better naming suggestions.

tiagoalvesdulce commented:
I decided to limit the proposal list loading skeleton a little bit because I think it improved the UX. LMK what you think. Relevant code:

```
const loadingPlaceholdersCount = max([
    min([inventory.length - records.length, 5]),
    0,
  ]);
```

1. Fixed by not showing the skeleton in all cases but just during loading state instead. This way the Intersection element can be visible at the end of the page.
2. Fixed by not calling `onRenderNextStatus` when it's not passed as a prop to the `RecordsStatusList` component (single status tabs)

victorgcramos reviewed (changes requested):
Great, empty lists are working as expected, and scrolling is way more consistent. But I actually found an issue on mobile screens, which seems like scrolling is not working, and does not trigger new fetches once I reach the end of page height.

Here's a preview:

![2769-proposals-scroll-mobile](https://user-images.githubusercontent.com/22639213/171700231-04c57367-1ba2-4fc3-9f3b-c5bfcf0b3951.gif)

victorgcramos reviewed (approved):
LGTM!
